src/toil/jobStores/aws/jobStore.py

```diff
@@ -74,38 +82,70 @@ def copyKeyMultipart(srcKey, dstBucketName, dstKeyName, headers=None):
+                                                         src_key_name=srcKey.name,
+                                                         src_version_id=srcKey.version_id,
+                                                         part_num=partIndex + 1,
+                                                         start=partIndex * partSize,
+                                                         end=min(partIndex * partSize + partSize,
+                                                                 totalSize) - 1,
+                                                         headers=headers)
+                log.debug('Successfully copied part %d of %d', partIndex, totalParts)
+
+            except Exception as e:
+                if len(exceptions) <= 4:
+                    exceptions[partIndex] = traceback.format_exc()
+                else:
+                    exceptions[partIndex] = e
+                log.debug('Part number %s of copy failed.', partIndex)
+                return None
+            else:
+                return part
+
     partSize = defaultPartSize
+    totalSize = srcKey.size
+    totalParts = (totalSize + partSize - 1) / partSize
+    exceptions = collections.OrderedDict()
     # We need a location-agnostic connection to S3 so we can't use the one that we
     # normally use for interacting with the job store bucket.
     with closing(boto.connect_s3()) as s3:
-        headers = headers or {}
-        totalSize = srcKey.size
         for attempt in retry_s3():
             with attempt:
                 dstBucket = s3.get_bucket(dstBucketName)
                 upload = dstBucket.initiate_multipart_upload(dstKeyName, headers=headers)
+        log.info("Initiated multipart copy from 's3://%s/%s' to 's3://%s/%s'.",
+                 srcKey.bucket.name, srcKey.name, dstBucketName, dstKeyName)
         try:
-            start = 0
-            partIndex = itertools.count()
-            while start < totalSize:
-                end = min(start + partSize, totalSize)
-                for attempt in retry_s3():
-                    with attempt:
-                        upload.copy_part_from_key(src_bucket_name=srcKey.bucket.name,
-                                                  src_key_name=srcKey.name,
-                                                  src_version_id=srcKey.version_id,
-                                                  part_num=next(partIndex) + 1,
-                                                  start=start,
-                                                  end=end - 1,
-                                                  headers=headers)
-                start += partSize
+            with ThreadPoolExecutor(max_workers=min(cpu_count() * 16, totalParts, 128)) as executor:
+                finishedParts = set(executor.map(copyPart, xrange(0, totalParts)))
+                if exceptions:
+                    for partIndex, tb in exceptions[0:4].items():
```

Comment:
You do not need to act on this, but I want you to know that using this slice here is still incorrect. The lesson we should learn is that we need better test coverage to expose these issues. Please study my upcoming commit to see how I dealt with it.